### ISSUE TITLE
Periodic `CoulombOperator` Gives Garbage

### ISSUE THREAD

**Issue by JonathonMisiewicz** [m-a-d-n-e-s-s/madness#535]
@evaleev and I are working on periodic code, and I find that values of the electric potential changed after rebasing onto a more recent version of MADNESS. The current results appear to have an additional "contribution" that is pure garbage (e.g., not even monotonic as you get away from the center of the sum-of-Gaussian charge distribution).

The problem is [this line of code](https://github.com/m-a-d-n-e-s-s/madness/blob/master/src/madness/mra/operator.h#L249) added during #498. The `CoulombOperator` required the last argument to `truncate_periodic_expansion` to be `true`, not `false`, [prior to the PR](https://github.com/m-a-d-n-e-s-s/madness/blob/b43c123d1e77169f1e305eebf1b0c1c8a5e0dee2/src/madness/mra/operator.h#L1628). Manually changing that argument reproduces our electric potential prior to rebasing onto #498, but doing that would break many other operators.

@fbischoff, I noticed that you generalized some operators in `operator.h` to use your new function but left others the same. Is there an underlying reason? I want to make sure that my fixes to the `CoulombOperator` don't interfere with anything you were trying to do.

**Comment by JonathonMisiewicz** [comment]
The periodic code is certainly dysfunctional for our purposes, and we're hoping to make it less so.

That helps. As a follow-up, why did `BSHOperator3D` get updated but not `BSHOperatorPtr3D`? Is there a reason for this, or was it simply a matter of convenience?

The trouble in our case is that Gaussians which should be discarded are not, and I'd also need to look into why they need to be discarded in some cases but not others... It could very well be a defect in the original code.

**Comment by fbischoff** [comment]
> The periodic code is certainly dysfunctional for our purposes, and we're hoping to make it less so.
I’m looking forward to that!

> That helps. As a follow-up, why did BSHOperator3D get updated but not BSHOperatorPtr3D? Is there a reason for this, or was it simply a matter of convenience?

no real reason, I tried to keep legacy code functional.

Best,
Florian



Florian Bischoff
Humboldt-Universitaet zu Berlin, Institut fuer Chemie
Unter den Linden 6, 10099 Berlin
phone: +49-30-2093-82708
office: Brook-Taylor-Str 2, room 3'322
***@***.***
https://tinyurl.com/bischofflab

**Comment by gifornl** [comment]
Did you set the function and operator to be periodic in the heading?

George

**Comment by JonathonMisiewicz** [comment]
Yes. The same application behaved _differently_ before/after #498, and I can restore the old behavior by restoring an argument flag to how it was pre-#498, so I doubt my application code is the issue.

The relevant part of our appliction code looks like
```
    BoundaryConditions<3> bc_periodic(BC_PERIODIC);
    FunctionDefaults<3>::set_bc(bc_periodic);

    auto r000 = std::vector<coord_3d>{coord_3d(std::array<double,3>{0.0, 0.0, 0.0})};
    // g*_000_func3d are standard functions to evaluate a gaussian
    Function<double, 3> gdiffuse_000 = FunctionFactory<double, 3>(world).initial_level(4).special_points(r000).f(gdiffuse_000_func3d);
    printf("projected gdiffuse_000\n");
    Function<double, 3> gtight_000 = FunctionFactory<double, 3>(world).initial_level(4).special_points(r000).f(gtight_000_func3d);
    auto gdiffuse = gdiffuse_000;
    auto gtight = gtight_000;
    auto rho = gdiffuse - gtight;

    SeparatedConvolution<double, 3> pop = CoulombOperator(world, 1e-10, eps, bc_periodic);
    Function<double, 3> V_periodic = apply(pop, rho);
```
where

**Comment by JonathonMisiewicz** [comment]
The picture is a little clearer now, so I'll try to clean this issue up soon.

**Comment by JonathonMisiewicz** [comment]
@fbischoff 

Another clarification regarding the recent refactor. Was the new constructor was added [for backwards-compatibility](https://github.com/m-a-d-n-e-s-s/madness/pull/498/files#diff-22333d357a4becfe78caa180594a5472c9adc7f780585e267719f78261b0c497R1007), so it therefore should be moved away from, or is that just an unfortunate copy-paste artifact?

I need to know what constructor to use when I'm repairing the operator construction.

**Comment by fbischoff** [comment]
Hi Jonathon, 

I’m not quite sure which constructor which copy-paste artifact you mean. My commit you’re referring to was quite (too?) large.. Maybe you can give me some specific line or so?

Best,
Florian


Florian Bischoff
Humboldt-Universitaet zu Berlin, Institut fuer Chemie
Unter den Linden 6, 10099 Berlin
phone: +49-30-2093-82708
office: Brook-Taylor-Str 2, room 3'322
***@***.***
https://tinyurl.com/bischofflab

> On Oct 9, 2024, at 5:42 PM, Jonathon Misiewicz ***@***.***> wrote:
> 
> 
> @fbischoff <https://github.com/fbischoff>
> Another clarification regarding the recent refactor. Was the new constructor was added, so it therefore should be moved away from, or is that just an unfortunate copy-paste artifact?
> 
https://github.com/m-a-d-n-e-s-s/madness/pull/498/files#diff-22333d357a4becfe78caa180594a5472c9adc7f780585e267719f78261b0c497R1007

**Comment by JonathonMisiewicz** [comment]
Ah, I didn't realize that GitHub's link-to-specific-line feature doesn't work for files with diffs too large to show...

[operator.h:1007](https://github.com/m-a-d-n-e-s-s/madness/pull/498/files#diff-22333d357a4becfe78caa180594a5472c9adc7f780585e267719f78261b0c497)

Note that the same comment is on line 1020.

**Comment by fbischoff** [comment]
Hi Jonathon,

I see. Comment line 1007 is a copy-paste artifact, sorry for that.. 

Best,
Florian



------------------------------------------------------
Florian Bischoff
Humboldt-Universitaet zu Berlin, Institut fuer Chemie
Unter den Linden 6, 10099 Berlin
phone: +49-30-2093-82708
office: Brook-Taylor-Str 2, room 3'322
***@***.***
https://tinyurl.com/bischofflab
